bmcs/src/main/scala/akka/stream/alpakka/oracle/bmcs/impl/BmcsStream.scala

```diff
@@ -0,0 +1,302 @@
+import akka.stream.alpakka.oracle.bmcs.scaladsl.ListObjectsResultContents
+import akka.stream.alpakka.s3.impl.{DiskBuffer, MemoryBuffer, SplitAfterSize}
+import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
+import akka.util.ByteString
+
+import scala.collection.immutable.Seq
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+sealed trait PartUploadResponse {
+  def multipartUpload: MultipartUpload
+
+  def index: Int
+}
+
+final case class SuccessfulPartUpload(multipartUpload: MultipartUpload, index: Int, etag: String)
+    extends PartUploadResponse
+
+final case class FailedPartUpload(multipartUpload: MultipartUpload, index: Int, exception: Throwable)
+    extends PartUploadResponse
+
+final case class FailedUpload(reasons: Seq[Throwable]) extends Exception(reasons.map(_.getMessage).mkString(", "))
+
+final case class CompleteMultipartUploadResult(bucket: String, objectName: String, etag: String)
+
+object BmcsStream {
+  def apply(settings: BmcsSettings, credentials: BmcsCredentials)(implicit system: ActorSystem,
+                                                                  mat: Materializer): BmcsStream =
+    new BmcsStream(settings, credentials)
+}
+
+class BmcsStream(settings: BmcsSettings, credentials: BmcsCredentials)(implicit system: ActorSystem, mat: Materializer)
+    extends BmcsJsonSupport {
+
+  import HttpRequests._
+
+  implicit val conf: BmcsSettings = settings
+
+  val MinChunkSize = 5242880 //in bytes
+
+  def listBucket(bucket: String, prefix: Option[String] = None): Source[ListObjectsResultContents, NotUsed] = {
+    sealed trait ListBucketState
+    case object Starting extends ListBucketState
+    case class Running(nextStart: String) extends ListBucketState
+    case object Finished extends ListBucketState
+
+    import system.dispatcher
+
+    def listBucketCall(start: Option[String]): Future[Option[(ListBucketState, Seq[ListObjectsResultContents])]] =
+      signAndGetAs[ListObjects](listObjectsRequest(bucket, start, prefix))
+        .map { (res: ListObjects) =>
+          val contents = res.objects.map(ListObjectsResultContents(_, bucket))
+          Some(
+            res.nextStartWith
+              .fold[(ListBucketState, Seq[ListObjectsResultContents])]((Finished, contents))(
+                n => (Running(n), contents)
+              )
+          )
+        }
+
+    Source
+      .unfoldAsync[ListBucketState, Seq[ListObjectsResultContents]](Starting) {
+        case Finished => Future.successful(None)
+        case Starting => listBucketCall(None)
+        case Running(st) => listBucketCall(Some(st))
+      }
+      .mapConcat(identity)
+  }
+
+  /**
+   *
+   * Returns a source of ByteString of the object in bmcs.
+   *
+   * @param bucket     the bucket
+   * @param objectName the object to download
+   * @param range      optional range of bytes.
+   * @return source built from the object in storage.
```

Comment:
"A source emitting the bytes of the requested object"?

Also good to document:
* Does the request happen when the returned source is materialized or right away?
* Can there returned source be materialized multiple times, 
* What happens on failure?

Reply:
updated comment.